File: r/decision_tree.R
# ===========================
# 🌳 Decision Tree Regression
# ===========================

# 📂 Import Dataset
dataset <- read.csv('../data/Position_Salaries.csv')
dataset <- dataset[, 2:3]  # Keep only Level and Salary columns

# 🧠 Fit Decision Tree Regression Model
library(rpart)
regressor <- rpart(
  formula = Salary ~ ., 
  data = dataset,
  control = rpart.control(minsplit = 2)
)

# 🔍 Predict Salary for Level = 6.5
y_pred <- predict(regressor, newdata = data.frame(Level = 6.5))
cat("💰 Predicted Salary for Level 6.5:", y_pred, "\n")

# ===========================
# 📊 Visualize Decision Tree Fit
# ===========================
library(ggplot2)

ggplot() +
  geom_point(
    aes(x = dataset$Level, y = dataset$Salary),
    color = "red"
  ) +
  geom_line(
    aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
    color = "blue"
  ) +
  labs(
    title = "Decision Tree Regression",
    x = "Level",
    y = "Salary"
  ) +
  theme_minimal()

# ===========================
# 📊 High-Resolution Plot
# ===========================
X_grid <- seq(min(dataset$Level), max(dataset$Level), 0.001)

ggplot() +
  geom_point(
    aes(x = dataset$Level, y = dataset$Salary),
    color = "red"
  ) +
  geom_line(
    aes(x = X_grid, y = predict(regressor, newdata = data.frame(Level = X_grid))),
    color = "blue"
  ) +
  labs(
    title = "Decision Tree Regression (High Resolution)",
    x = "Level",
    y = "Salary"
  ) +
  theme_minimal()

File: python/decision_tree.py
# ===========================
# 🌳 Decision Tree Regression
# ===========================

# 📦 Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeRegressor

# 📥 Load Dataset
dataset = pd.read_csv('../data/Position_Salaries.csv')
dataset = dataset.iloc[:, 1:]  # Extract Level and Salary columns
X = dataset.iloc[:, [0]].values
y = dataset.iloc[:, -1].values

# 🧠 Fit the Decision Tree model
regressor = DecisionTreeRegressor(random_state=0)
regressor.fit(X, y)

# 🔍 Predict the salary for level 6.5
y_pred = regressor.predict([[6.5]])
print(f"💰 Predicted Salary for level 6.5: ${y_pred[0]:,.2f}")

# ===========================
# 📊 Visualize Decision Tree Fit
# ===========================

# 🔵 Basic scatter + regression
plt.scatter(X, y, color='red', label='Actual')
plt.plot(X, regressor.predict(X), color='blue', label='Predicted')
plt.title('Decision Tree Regression')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.grid(True)
plt.legend()
plt.tight_layout()
plt.show()

# 🔵 High-resolution curve
X_grid = np.arange(X.min(), X.max(), 0.01).reshape(-1, 1)  # Finer resolution
plt.scatter(X, y, color='red', label='Actual')
plt.plot(X_grid, regressor.predict(X_grid), color='blue', label='Predicted')
plt.title('Decision Tree Regression (High Resolution)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.grid(True)
plt.legend()
plt.tight_layout()
plt.show()

# ✅ Done
print("✅ Decision Tree Regression complete. Plots displayed successfully!")
